app/command_invite.go

```diff
@@ -0,0 +1,104 @@
+	return CMD_INVITE
+}
+
+func (me *InviteProvider) GetCommand(a *App, T goi18n.TranslateFunc) *model.Command {
+	return &model.Command{
+		Trigger:          CMD_INVITE,
+		AutoComplete:     true,
+		AutoCompleteDesc: T("api.command_invite.desc"),
+		AutoCompleteHint: T("api.command_invite.hint"),
+		DisplayName:      T("api.command_invite.name"),
+	}
+}
+
+func (me *InviteProvider) DoCommand(a *App, args *model.CommandArgs, message string) *model.CommandResponse {
+	if message == "" {
+		return &model.CommandResponse{Text: args.T("api.command_invite.missing_message.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	}
+
+	l4g.Debug(message)
+
+	splitMessage := strings.SplitN(message, " ", 2)
+	targetUsername := splitMessage[0]
+	targetUsername = strings.TrimPrefix(targetUsername, "@")
+
+	var userProfile *model.User
+	if result := <-a.Srv.Store.User().GetByUsername(targetUsername); result.Err != nil {
+		l4g.Error(result.Err.Error())
+		return &model.CommandResponse{Text: args.T("api.command_invite.missing_user.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	} else {
+		userProfile = result.Data.(*model.User)
+	}
+
+	var channelToJoin *model.Channel
+	var err *model.AppError
+	// User set a channel to add the invited user
+	if len(splitMessage) > 1 && splitMessage[1] != "" {
+		targetChannelName := strings.TrimPrefix(strings.TrimSpace(splitMessage[1]), "~")
+
+		if channelToJoin, err = a.GetChannelByName(targetChannelName, args.TeamId); err != nil {
+			return &model.CommandResponse{Text: args.T("api.command_invite.channel.error", map[string]interface{}{"Channel": targetChannelName}), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+		}
+	} else {
+		channelToJoin, err = a.GetChannel(args.ChannelId)
+		if err != nil {
+			return &model.CommandResponse{Text: args.T("api.command_invite.channel.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+		}
+	}
+
+	// Check if is a Direct Channel
+	if channelToJoin.Type == model.CHANNEL_DIRECT || channelToJoin.Type == model.CHANNEL_GROUP {
+		return &model.CommandResponse{Text: args.T("api.command_invite.directchannel.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	}
+
+	// Check if user is already in the channel
+	userChannels, err := a.GetChannelsForUser(args.TeamId, userProfile.Id)
+	if err != nil {
+		return &model.CommandResponse{Text: args.T("api.command_invite.fail_get_userchannel.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	}
+
+	for _, channel := range *userChannels {
+		if channel.Id == channelToJoin.Id {
+			return &model.CommandResponse{Text: args.T("api.command_invite.user_already_in_channel.app_error", map[string]interface{}{"User": userProfile.Username}), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+		}
+	}
+
+	if channelToJoin.Type == model.CHANNEL_PRIVATE && !a.SessionHasPermissionToChannel(args.Session, channelToJoin.Id, model.PERMISSION_READ_CHANNEL) {
```

Comment:
The permissions checking would be the same that used in `api4/channel.go` `addChannelMember` function. It is a bit more complex (depends on if you are adding yourself, or other, and if you have the permissions to manage the team members).

the code in the other file is:
```go
    if channel.Type == model.CHANNEL_OPEN {
        if member.UserId == c.Session.UserId {
            if !c.App.SessionHasPermissionToChannel(c.Session, channel.Id, model.PERMISSION_JOIN_PUBLIC_CHANNELS) {
                c.SetPermissionError(model.PERMISSION_JOIN_PUBLIC_CHANNELS)
                return
            }
        } else {
            if !c.App.SessionHasPermissionToChannel(c.Session, channel.Id, model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS) {
                c.SetPermissionError(model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS)
                return
            }
        }
    }

    if channel.Type == model.CHANNEL_PRIVATE && !c.App.SessionHasPermissionToChannel(c.Session, channel.Id, model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS) {
        c.SetPermissionError(model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS)
        return
    }
```

Maybe in this case we can assume that you are trying to invite another person, so you will never invite yourself to channels.

Reply:
done! thanks so much for your review. PTAL